### PR TITLE
Add diagnostic logs at startup

### PR DESCRIPTION
**Pull request by kevingosse**

### CODE REVIEW

File: src/Datadog.Trace/Tracer.cs [hunk 1]
```diff
@@ -415,6 +417,91 @@ internal void StartDiagnosticObserversInternal()
...
+
+                    writer.WritePropertyName("lang");
+                    writer.WriteValue(".net");
+
+                    writer.WritePropertyName("lang_version");
+#if NETSTANDARD
+                    writer.WriteValue(System.Runtime.InteropServices.RuntimeInformation.FrameworkDescription);
+#else
+                    writer.WriteValue(Environment.Version.ToString());
+#endif
```
Comment (lucaspimentel) [inline src/Datadog.Trace/Tracer.cs:451]: We have a `FrameworkDescription` helper class which takes care of some problems with using `Environment.Version`.

File: src/Datadog.Trace/Tracer.cs [hunk 2]
```diff
@@ -415,6 +417,91 @@ internal void StartDiagnosticObserversInternal()
...
+
+                    writer.WritePropertyName("date");
+                    writer.WriteValue(DateTime.Now);
+
+                    writer.WritePropertyName("os_name");
+#if NETSTANDARD
+                    writer.WriteValue(System.Runtime.InteropServices.RuntimeInformation.OSDescription);
+#else
+                    writer.WriteValue(Environment.OSVersion.ToString());
+#endif
```
Comment (lucaspimentel) [inline src/Datadog.Trace/Tracer.cs:438]: We should probably include OS platform (x64, x86, etc) and process platform. We should probably get this form `FrameworkDescription` to keep it consistent with other places where we report this.

Comment (kevingosse) [reply src/Datadog.Trace/Tracer.cs:438]: Good point, gonna add the process platform. Is there any value in including the OS platform?

File: src/Datadog.Trace/Tracer.cs (hunk 1 above)
Comment (kevingosse) [reply src/Datadog.Trace/Tracer.cs:451]: Nice!

File: src/Datadog.Trace/Tracer.cs (hunk 2 above)
Comment (lucaspimentel) [reply src/Datadog.Trace/Tracer.cs]: Hmm, I guess not. Never mind!

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 3]
```diff
@@ -183,7 +183,7 @@ CorProfiler::Initialize(IUnknown* cor_profiler_info_unknown) {
     }
   }
   if (FAILED(hr)) {
-    Warn("Failed to attach profiler: unable to set event mask.");
+    Warn("DATADOG TRACER DIAGNOSTICS - Failed to attach profiler: unable to set event mask.");
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:186]: I'm not sure what qualifies for `DATADOG TRACER DIAGNOSTICS`, but do the above domain-neutral logs need this message prefix or no?

File: src/Datadog.Trace/Tracer.cs [hunk 4]
```diff
@@ -415,6 +427,123 @@ internal void StartDiagnosticObserversInternal()
             }
         }
 
+        internal async Task WriteDiagnosticLog()
```
Comment (zacharycmontoya) [inline src/Datadog.Trace/Tracer.cs:430]: In Azure App Services we may initialize the Tracer to send to one port for the trace-agent and one port for dogstatsd, and then update it later after using the filesystem to record which ports we were able to open for those processes. Perhaps this `WriteDiagnosticLog` method should be included in the `TracingProcessManager.SubscribeToTraceAgentPortOverride` and `TracingProcessManager.SubscribeToDogStatsDPortOverride` callbacks, if that won't be too noisy.

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp (hunk 3 above)
Comment (kevingosse) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:186]: As I understand it, `DATADOG TRACER DIAGNOSTICS` should be used for stuff that causes the instrumentation to fail. The domain-neutral logs does not necessarily indicate an error condition.

File: src/Datadog.Trace/Tracer.cs (hunk 4 above)
Comment (kevingosse) [reply src/Datadog.Trace/Tracer.cs:430]: The issue is that we're supposed to have only one "DATADOG TRACER CONFIGURATION" log. I'll leave it as is for now.

**Pull request merged**